networkx/algorithms/operators/product.py

```diff
@@ -532,3 +533,94 @@ def corona_product(G, H):
         GH.add_edges_from((G_node, (G_node, H_node)) for H_node in H)
 
     return GH
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+@nx._dispatchable(graphs=_G_H)
+def modular_product(G, H):
+    r"""Returns the Modular product of G and H.
+
+    The modular product of $G$ and $H$ is the graph $M = G \nabla H$,
+    consisting of the node set $V(M) = V(G) \times V(H)$ that is the Cartesian
+    product of the node sets of $G$ and $H$. Further, $M$ contains an edge $((u, v), (x, y))$:
+
+    - if $u$ is adjacent to $x$ in $G$ and $v$ is adjacent to $y$ in $H$, or
+    - if $u$ is not adjacent to $x$ in $G$ and $v$ is not adjacent to $y$ in $H$.
+
+    More formally
+    ::
+
+        E(M) = {((u, v), (x, y)) | ((u, x) in E(G) and (v, y) in E(H)) or
+                                   ((u, x) not in E(G) and (v, y) not in E(H))}
+    Parameters
+    ----------
+    G, H: NetworkX graphs
+        The graphs to take the modular product of.
+
+    Returns
+    -------
+    M: NetworkX graph
+        The Modular product of G and H.
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If G is not a simple graph.
+
+    Examples
+    --------
+    >>> G = nx.cycle_graph(4)
+    >>> H = nx.path_graph(2)
+    >>> M = nx.modular_product(G, H)
+    >>> list(M)
+    [(0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1), (3, 0), (3, 1)]
+    >>> print(M)
+    Graph with 8 nodes and 8 edges
+
+    Notes
+    -----
+    The *modular product* is defined in [1]_ and was first
+    introduced as the *weak modular product*.
+
+    The modular product reduces the problem of counting isomorphic subgraphs
+    in $G$ and $H$ to the problem of counting cliques in $M$. The subgraphs of
+    $G$ and $H$ that are induced by the nodes of a clique in $M$ are
+    isomorphic [2]_ [3]_.
+
+    References
+    ----------
+    .. [1] R. Hammack, W. Imrich, and S. Klavžar,
+        "Handbook of Product Graphs", CRC Press, 2011.
+
+    .. [2] H. G. Barrow and R. M. Burstall,
+        "Subgraph isomorphism, matching relational structures and maximal
+        cliques", Information Processing Letters, vol. 4, issue 4, pp. 83-84,
+        1976, https://doi.org/10.1016/0020-0190(76)90049-1.
+
+    .. [3] V. G. Vizing, "Reduction of the problem of isomorphism and isomorphic
+        entrance to the task of finding the nondensity of a graph." Proc. Third
+        All-Union Conference on Problems of Theoretical Cybernetics. 1974.
+    """
+    if H.is_directed():
+        raise nx.NetworkXNotImplemented("Modular not implemented for directed graphs")
+    if H.is_multigraph():
+        raise nx.NetworkXNotImplemented("Modular not implemented for multigraphs")
```

Comment:
I am also very unhappy with these checks, as they seem inappropriate at this position in the code, however afaik it is not correct, that this is checked by the `@not_implemented_for` decorator. I looked into it a little bit, and it seems as if only the first parameter is decorated. Therefore, the checks for graph-type that are currently tested do not work without the checks. Specifically, the tests, where the first parameter is a simple graph and the second is directed or multi break:
```

============================================================================ FAILURES =============================================================================
___________________________________________________________________ test_modular_product_raises ___________________________________________________________________

    def test_modular_product_raises():
        G = nx.Graph([(0, 1), (1, 2), (2, 0)])
        H = nx.Graph([(0, 1), (1, 2), (2, 0)])
        DG = nx.DiGraph([(0, 1), (1, 2), (2, 0)])
        DH = nx.DiGraph([(0, 1), (1, 2), (2, 0)])
        with pytest.raises(nx.NetworkXNotImplemented):
>           nx.modular_product(G, DH)

networkx/algorithms/operators/tests/test_product.py:475: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
<class 'networkx.utils.decorators.argmap'> compilation 43:5: in argmap_modular_product_39
    ???
networkx/utils/backends.py:550: in __call__
    return self.orig_func(*args, **kwargs)
networkx/algorithms/operators/product.py:606: in modular_product
    GH = _init_product_graph(G, H)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

G = <networkx.classes.graph.Graph object at 0x7f4ada688670>, H = <networkx.classes.digraph.DiGraph object at 0x7f4ada689960>

    def _init_product_graph(G, H):
        if G.is_directed() != H.is_directed():
            msg = "G and H must be both directed or both undirected"
>           raise nx.NetworkXError(msg)
E           networkx.exception.NetworkXError: G and H must be both directed or both undirected

networkx/algorithms/operators/product.py:117: NetworkXError
===================================================================== short test summary info =====================================================================
FAILED networkx/algorithms/operators/tests/test_product.py::test_modular_product_raises - networkx.exception.NetworkXError: G and H must be both directed or both undirected
================================================================== 1 failed, 29 passed in 0.17s ===================================================================
```

Maybe I do not use the annotations correctly, yet I tried to follow the documentation of the decorator.

The options I see are:
- Leave it as is
-  Use the check in `_init_product_graph(G, H)` which results in a `NetworkXError`, not a `NetworkXNotImplemented`.
- Implement a similar decorator for operators that allows to check parameters individually for type.

Reply:
Ah - good point, I missed that subtlety. There may be a clever way to leverage `argmap` to do this (@boothby ), but let's leave that for a follow-up PR!

My vote for now would be to remove the decorators then and just make the checks explicit in the code:

```suggestion
    if G.is_directed() or H.is_directed():
        raise nx.NetworkXNotImplemented("Modular product not implemented for directed graphs")
    if G.is_multigraph() or H.is_multigraph():
        raise nx.NetworkXNotImplemented("Modular product not implemented for multigraphs")
```

The `not_implemented_for` decorator is supposed to help cut down on boilerplate, but since we need to check both arguments I think it's more readable if all of the checks are explicit and in the same location, rather than split between the decorator and the code.

None of this is really a blocker though!

BTW - great job with the tests, it's always nice when tests are so well-thought out that they catch subtleties like this. Bravo!

Reply:
You are right as usual @rossbar!  The argmap decorator does allow multiple inputs to be decorated. The fix here needs to be in `not_implemented_for` where the call is current `argmap(_not_implement_for, 0)`.  We should be able to change the `0` to be an optional input to the decorator.

There may be quite a few places in the code where this would be an opportunity -- that is where there are more than one graph input, or where the graph input is not the first argument.